File: node/sample_07_market_order.js
/**
 * Sample 7 — Dat lenh Market (MTL)
 * ===================================
 * Khop lenh nhanh theo gia thi truong hien tai.
 *
 * Luong:
 *   1. Client tao order MARKET (khong gui price)
 *   2. Gui request toi Trading Orders API
 *   3. He thong match theo thanh khoan thi truong tai thoi diem gui
 *   4. API tra ve trang thai khop (FILLED hoac PARTIAL_FILLED)
 *   5. Cap nhat ngay danh muc/so du tam tinh
 */

import { Auth, Trading, OrderSide } from '@ssi.developer/ssi-sdk';
import { config, ACCOUNT_NO, OTP } from './config.js';
import { ensureAuth } from './auth_helper.js';

const vnd = (x) => Number(x).toLocaleString('en-US', { maximumFractionDigits: 0 });

const auth = new Auth(config);
await ensureAuth(auth, OTP);

const trading = new Trading(auth);

// --- Buoc 1: Kiem tra suc mua/ban o gia thi truong ---
const maxBs = await trading.trading.getMaxBuySellAtMarketPrice(ACCOUNT_NO, 'SSI');
console.log(`Max mua (market): ${maxBs.maxBuyQuantity} co phieu`);
console.log(`Max ban (market): ${maxBs.maxSellQuantity} co phieu`);

// --- Buoc 2: Dat lenh Market mua ---
console.log('\n--- Dat lenh MARKET mua SSI ---');
const result = await trading.trading.placeMarketOrder(ACCOUNT_NO, 'SSI', OrderSide.BUY, 100);
console.log('  Ket qua:', result);

// --- Buoc 3: Kiem tra trang thai lenh ---
console.log('\n--- So lenh hom nay ---');
const orders = await trading.portfolio.getTodayOrders(ACCOUNT_NO);
for (const order of orders.slice(-3)) {
  console.log(
    `  ${order.orderId} | ${order.symbol} ${order.side} ${order.orderType} | ` +
    `SL: ${order.quantity} | Khop: ${order.filledQuantity} | Trang thai: ${order.status}`,
  );
}

// --- Buoc 4: Cap nhat lai so du sau khi khop ---
console.log('\n--- So du sau giao dich ---');
const balance = await trading.portfolio.getEquityBalance(ACCOUNT_NO);
console.log(`  Tien mat kha dung: ${vnd(balance.availableCash).padStart(15)}`);

// --- Buoc 5: Cap nhat danh muc ---
console.log('\n--- Vi the sau giao dich ---');
const positions = await trading.portfolio.getEquityPositions(ACCOUNT_NO);
for (const pos of positions) {
  if (pos.symbol === 'SSI') {
    console.log(`  SSI | SL: ${pos.quantity} | Gia von: ${vnd(pos.costPrice)}`);
  }
}

// --- Response Summary ---
console.log(`\n[Response] max_buy_mkt|max_sell_mkt|buy_status`);
console.log(`${maxBs.maxBuyQuantity}|${maxBs.maxSellQuantity}|${result.status}`);

File: python/sample_07_market_order.py
"""
Sample 7 — Đặt lệnh Market (MP)
==================================
Khớp lệnh nhanh theo giá thị trường hiện tại.

Luồng:
  1. Client tạo order MARKET (không gửi price)
  2. Gửi request tới Trading Orders API
  3. Hệ thống match theo thanh khoản thị trường tại thời điểm gửi
  4. API trả về trạng thái khớp (FILLED hoặc PARTIALLY_FILLED)
  5. Cập nhật ngay danh mục/số dư tạm tính
"""

from ssi_sdk import Auth, Trading
from auth_helper import ensure_auth
from ssi_sdk.enums import OrderSide
from config import config, ACCOUNT_NO, OTP

with Auth(config) as auth:
    ensure_auth(auth, otp=OTP)

    with Trading(auth) as trading:
        # --- Bước 1: Kiểm tra sức mua/bán ở giá thị trường ---
        max_bs = trading.trading.get_max_buy_sell_at_market_price(ACCOUNT_NO, "SSI")
        print(f"Max mua (market): {max_bs.max_buy_quantity} cổ phiếu")
        print(f"Max bán (market): {max_bs.max_sell_quantity} cổ phiếu")

        # --- Bước 2: Đặt lệnh Market mua ---
        print("\n--- Đặt lệnh MARKET mua SSI ---")
        result = trading.trading.place_market_order(
            account_no=ACCOUNT_NO,
            symbol="SSI",
            side=OrderSide.BUY,
            quantity=100,
        )
        print(f"  Kết quả: {result}")

        # --- Bước 3: Kiểm tra trạng thái lệnh ---
        print("\n--- Sổ lệnh hôm nay ---")
        orders = trading.portfolio.get_today_orders(ACCOUNT_NO)
        for order in orders[-3:]:
            print(
                f"  {order.order_id} | {order.symbol} {order.side.value} "
                f"{order.order_type.value} | SL: {order.quantity} "
                f"| Khớp: {order.filled_quantity} | Trạng thái: {order.status.value}"
            )

        # --- Bước 4: Cập nhật lại số dư sau khi khớp ---
        print("\n--- Số dư sau giao dịch ---")
        balance = trading.portfolio.get_equity_balance(ACCOUNT_NO)
        print(f"  Tiền mặt khả dụng: {balance.account_balance:>15,.0f}")

    # --- Bước 5: Cập nhật danh mục ---
    print("\n--- Vị thế sau giao dịch ---")
    positions = trading.portfolio.get_equity_positions(ACCOUNT_NO)
    for pos in positions:
        if pos.symbol == "SSI":
            print(f"  SSI | SL: {pos.quantity} | Giá vốn: {pos.cost_price:,.0f}")
